File: core/ai/src/commonMain/kotlin/dev/audiocompanion/ai/PersonalContextStore.kt
package dev.audiocompanion.ai

import kotlinx.io.buffered
import kotlinx.io.files.FileSystem
import kotlinx.io.files.Path
import kotlinx.io.readByteArray
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Durable store for [PersonalContext] at `<root>/ai/personal_context.json`, written via temp file +
 * atomic rename like other stores in this app.
 */
class FilePersonalContextStore(
    private val fileSystem: FileSystem,
    root: Path,
) {
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
    }
    private val aiDir = Path(root, "ai")
    private val contextPath = Path(aiDir, FILE_NAME)

    fun load(): PersonalContext {
        if (!fileSystem.exists(contextPath)) return PersonalContext()
        val text = fileSystem.source(contextPath).buffered().use { it.readByteArray() }.decodeToString()
        return runCatching { json.decodeFromString(PersonalContext.serializer(), text) }
            .getOrDefault(PersonalContext())
    }

    fun save(context: PersonalContext): PersonalContext {
        write(context)
        return context
    }

    fun clear() {
        fileSystem.delete(contextPath, mustExist = false)
        fileSystem.delete(Path(aiDir, "$FILE_NAME.tmp"), mustExist = false)
    }

    private fun write(context: PersonalContext) {
        fileSystem.createDirectories(aiDir)
        val tmp = Path(aiDir, "$FILE_NAME.tmp")
        fileSystem.sink(tmp).buffered().use { sink ->
            sink.write(json.encodeToString(PersonalContext.serializer(), context).encodeToByteArray())
        }
        fileSystem.atomicMove(tmp, contextPath)
    }

    companion object {
        const val FILE_NAME = "personal_context.json"
    }
}

File: ios/PebbleAudioKit/Sources/Intelligence/PersonalContextStore.swift
import Foundation

// Port of `core/ai/.../PersonalContextStore.kt`.

/// Durable store for `PersonalContext` at `<root>/ai/personal_context.json`, written via temp
/// file + atomic rename like other stores in this app. The JSON shape matches the KMP file
/// exactly (see PersonalContext.swift) — this is user-authored data.
public final class FilePersonalContextStore: @unchecked Sendable {
    public static let fileName = "personal_context.json"

    private let aiDir: URL
    private let contextPath: URL

    public init(root: URL) {
        self.aiDir = root.appendingPathComponent("ai", isDirectory: true)
        self.contextPath = aiDir.appendingPathComponent(Self.fileName)
    }

    public func load() -> PersonalContext {
        guard let data = try? Data(contentsOf: contextPath) else { return PersonalContext() }
        return (try? JSONDecoder().decode(PersonalContext.self, from: data)) ?? PersonalContext()
    }

    @discardableResult
    public func save(_ context: PersonalContext) throws -> PersonalContext {
        try write(context)
        return context
    }

    public func clear() throws {
        try removeIfExists(contextPath)
        try removeIfExists(aiDir.appendingPathComponent("\(Self.fileName).tmp"))
    }

    private func write(_ context: PersonalContext) throws {
        try FileManager.default.createDirectory(at: aiDir, withIntermediateDirectories: true)
        let tmp = aiDir.appendingPathComponent("\(Self.fileName).tmp")
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(context).write(to: tmp)
        try atomicMove(from: tmp, to: contextPath)
    }

    private func removeIfExists(_ url: URL) throws {
        guard FileManager.default.fileExists(atPath: url.path) else { return }
        try FileManager.default.removeItem(at: url)
    }

    /// POSIX rename: atomically replaces `to` (Kotlin `FileSystem.atomicMove` semantics).
    private func atomicMove(from: URL, to: URL) throws {
        if rename(from.path, to.path) != 0 {
            throw NSError(
                domain: NSPOSIXErrorDomain, code: Int(errno),
                userInfo: [NSFilePathErrorKey: to.path])
        }
    }
}
